### ISSUE TITLE
When entering a map title it should accept spaces and replace them with hyphens in the URL

### ISSUE THREAD

**Issue by prototyperspective** [amelioro/ameliorate#536]
### Describe your issue

When creating a new topic most users will probably add a sentence with spaces. It will show an error and many will not understand what the problem is. Moreover, it shouldn't be a problem at all.

### Solution you'd like

The title with spaces could be displayed somewhere at the map and the URL get automatically changed so it replaces spaces with hyphens -

### Additional context

Created a topic named like https://ameliorate.app/Prototyperspective/CarsAreHarmfulToSociety because I didn't think of using spaces because I didn't know this title was used for the URL. It would be better to have all topics with spaces like …/problems-of-cars

**Comment by keyserj** [comment]
Yeah this seems good to do at some point. I think these are roughly the needed changes:
1. allow the `topic.title` column to have spaces (currently is disallowed via zod validation)
2. continue displaying the title wherever we currently are
3. replace all spaces with hyphens wherever the title is used for a URL

Known places where title is used for URL:
1. topic settings "view your topic at..."
2. user page, topic column (display text should be same, anchor text should be replaced)
3. topic details pane "[username]/[title]" (display text should be same, anchor text should be replaced)

**Comment by lovelypangotra2003** [comment]
@keyserj  Can I try this one, please?

**Comment by keyserj** [comment]
@lovelypangotra2003 yep assigned! Let me know if you have questions.

**Comment by lovelypangotra2003** [comment]
While working on this I got the following error `Message: ERROR: nondeterministic collations are not supported for substring searches`. Would it be acceptable to update the schema to handle this, or would you prefer a different approach?
Also, regarding transforming the title into a hyphenated URL. Would you recommend creating a reusable utility function (similar to what was mentioned in #356) or should I simply use `.replace(...)` on `topic.title` wherever needed?

**Comment by keyserj** [comment]
@lovelypangotra2003 regarding that error, what does your query look like? `substring searches` makes it seem like you may be using the `LIKE` operator but my expectation is that we can use `=` with `REPLACE`.

We're using a [nondeterministic collation](https://www.postgresql.org/docs/current/collation.html#COLLATION-NONDETERMINISTIC) so that the title column is [case insensitive for comparisons](https://stackoverflow.com/a/70742041/8409296). I hadn't realized this came with the limitation on substring searches, TIL! Potentially this could be changed, but only if such a change is necessary (we'd have to update existing queries to become case insensitive).

> regarding transforming the title into a hyphenated URL. Would you recommend creating a reusable utility function (similar to what was mentioned in https://github.com/amelioro/ameliorate/issues/356) or should I simply use .replace(...) on topic.title wherever needed?

Good callout. We've already got a method for [getLinkToTopic](https://github.com/amelioro/ameliorate/blob/7d9424b70256c90f1f13aa309844d2ba5e2b024d/src/common/topic.ts#L36) which seems like a good, single place for the logic. If there are changes needed elsewhere, maybe those places could be refactored to use this one?

**Comment by lovelypangotra2003** [comment]
@keyserj I was using `=` with `REPLACE`, when I got that error. If you have any direction I could proceed this in that'll be great, I'll keep trying various other ways till then.

**Comment by keyserj** [comment]
@lovelypangotra2003 can you paste in the exact query you're running? I've just tried running a `= REPLACE` query locally and I don't get the error:
```sql
SELECT * FROM public.topics
WHERE title = REPLACE('cars going too fast', ' ', '-')
```

**Comment by keyserj** [comment]
@lovelypangotra2003 woops, that makes sense 🤦😄. 

What do you mean by "creating a separate slug field"? Like, one column for `title` that allows anything, like "cars going-too_fast" and another for `titleDashes` that is just `title` converted to be consistent for searching without REPLACE, like "cars-going-too-fast"? I think that could work, but maintaining a separate column sounds a bit annoying.

I'm thinking that it'd be best to change the schema to use `citext` for case insensitivity rather than rely on collation, so that we can rely on `REPLACE` rather than creating another column. After brushing up on [Case-insensitive pattern matching in Postgresql](https://www.cybertec-postgresql.com/en/case-insensitive-pattern-matching-in-postgresql/), here are some note on the options:
1. case-insensitive collation or `citext` column type are both good because they ensure the column is unique across casings without needing to be explicit about it for each query
2. collation is nice because it can go with a `varchar(100)` column type to ensure a max length at the database level. `citext` would need a [check constraint](https://stackoverflow.com/questions/29560016/how-to-add-a-length-constraint-to-a-text-field) to validate this at the database level. But I actually don't think having this at the database-level is necessary, because [we already have a common zod schema that validates max length](https://github.com/amelioro/ameliorate/blob/7d9424b70256c90f1f13aa309844d2ba5e2b024d/src/common/topic.ts#L19) and that is easily (and already) used at the API-level.
3. nondeterministic collation (e.g. this case insensitive one) does not support pattern matching, but `citext` does.
4. `citext` has lower performance for longer strings because it internally converts all characters to lowercase, but our column will have max length of 100 which isn't very long.
5. it does seem like postgres v18 will support pattern matching with nondeterministic collations (we're on v15), but v18 is still in development, and it doesn't seem worth waiting for that.

So all that is to say, I think we can do this:
1. change the `title` column to use `TYPE CITEXT COLLATE pg_catalog."default"` instead of the current `TYPE VARCHAR(100) COLLATE english_ci;` (`english_ci` is the nondeterministic collation)
2. to use the `citext` type we'll also need to enable the `citext` postgres extension. It looks like we can [update the prisma schema](https://www.prisma.io/docs/orm/prisma-schema/data-model/unsupported-database-features#enable-postgresql-extensions-for-native-database-functions) then generate a migration from that.
3. then we can rely on `REPLACE` for our queries on `title`.

What do you think? This would require you to deal with prisma migrations now, which will be another thing to learn if you're not familiar already. See [CONTRIBUTING#database](https://github.com/amelioro/ameliorate/blob/main/CONTRIBUTING.md#database) for some notes on working with the database for Ameliorate.

**Comment by lovelypangotra2003** [comment]
Hey @keyserj , I was able to do all the things that you mentioned but now I ran into another conflict.
Right now, we're overloading the title field to serve both purposes — but this creates issues:

I’ve run into a core issue while implementing support for spaces in topic titles. Here’s what’s happening:

When a user creates a topic like `new topic`, the frontend correctly converts it to a URL-safe version like `/new-topic`.
On the backend, I normalize this back to `new topic` to match it against the stored title (since we currently store the original title with spaces). This works fine.
However, the issue appears when a user explicitly creates a topic like `climate-crisis` (with actual hyphens in the title):

The backend normalization replaces the hyphens with spaces (e.g., `climate-crisis` becomes "climate crisis"), and then it fails to find the topic, resulting in a 404.
So I’m currently stuck between:

Normalizing the title (which breaks legitimate hyphenated titles), or
Not normalizing it (which breaks titles with spaces).

From what I found, there’s no reliable way to distinguish a user who meant to use spaces vs. hyphens, since we're overloading the title field for both display and URL identity and the only solution around this is again by creating another column.

The column `title` can be used for titles like `new topic`
The new column can be used for titles like `new-topic`

Let me know if this should be implemented or if you have any other solutions.

**Comment by keyserj** [comment]
@lovelypangotra2003 This is what I'm thinking:
- the backend stores exactly what the user entered, no converting between spaces/dashes (because, as you said, there's no other way to know if the user meant to use spaces or dashes)
- the `topic.findByUsernameAndTitle` endpoint returns the backend value (so that the user-specified value can be displayed), _but_ the query for retrieving the topic normalizes to dashes to ensure we always find the same topic across space/dash variations (this should also ensure uniqueness)
- wherever we build the URL, we normalize to dashes for URL safety

Do you see any issues with this strategy?

**Comment by keyserj** [comment]
@lovelypangotra2003 it's been a while so I'm going to unassign you, let me know if you want to be reassigned! Also I really appreciate the work you put into this, I'll be able to finish it myself pretty easily I think (assuming you don't come back to finish). 🙏
